packages/melody-hooks/src/component.js

```diff
@@ -0,0 +1,181 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { enqueueComponent, options } from 'melody-idom';
+import { shallowEquals, shallowEqualsScalar } from './util/shallowEquals';
+import { HOOK_TYPE_USE_EFFECT, HOOK_TYPE_USE_REF } from './constants';
+import { setCurrentComponent, unsetCurrentComponent } from './util/hooks';
+
+const defaultComponentFn = props => props;
+
+function Component(element, componentFn) {
+    // part of the public API
+    this.el = element;
+
+    // needed for this type of component
+    this.componentFn = componentFn ? componentFn : defaultComponentFn;
+    this.hooks = [];
+    this.hooksPointer = -1;
+    this.isMounted = false;
+    this.props = null;
+    this.state = {};
+    this.data = null;
+
+    this.setState = this.setState.bind(this);
+}
+
+Object.assign(Component.prototype, {
+    /**
+     * Set new properties for the Component.
+     * This might cause the component to request an update.
+     */
+    apply(props) {
+        if (shallowEquals(props, this.props)) {
+            return;
+        }
+        this.props = props || null;
+        this.enqueueComponent();
+    },
+
+    /**
+     * Executed after a component has been mounted or updated.
+     * After this method has been triggered, the component is considered stable and
+     * accessing the DOM should be safe.
+     * The children of this Component might not have been rendered.
+     */
+    notify() {
+        const hooks = this.hooks;
+        for (let i = 0, l = hooks.length; i < l; i++) {
+            const hook = hooks[i];
+            const type = hook[0];
+
+            switch (type) {
+                case HOOK_TYPE_USE_EFFECT: {
+                    const dirty = hook[3];
+                    if (!dirty) continue;
+
+                    const callback = hook[1];
+                    const unsubscribe = hook[4];
+
+                    if (unsubscribe) unsubscribe();
+                    const unsubscribeNext = callback() || null;
+                    hook[4] = unsubscribeNext;
+                    break;
+                }
+                default: {
+                    continue;
+                }
+            }
+        }
+
+        if (this.isMounted) {
+            if (options.afterUpdate) {
+                options.afterUpdate(this);
+            }
+        } else {
+            this.isMounted = true;
+            if (options.afterMount) {
+                options.afterMount(this);
+            }
+        }
+    },
+
+    setState(hookIndex, valueNext) {
+        const value = this.state[hookIndex];
+        if (shallowEqualsScalar(value, valueNext)) {
+            return;
+        }
+        this.state[hookIndex] = valueNext;
+        this.enqueueComponent();
+    },
+
+    enqueueComponent() {
+        setCurrentComponent(this);
+        this.hooksPointer = -1;
+        this.data = this.componentFn(this.props) || {};
```

Comment:
I see a potential infinite loop:
`componentFn -> useState -> setState -> enqueueComponent`
Maybe we can prevent users from calling state setters while `componentFn` is running.

Reply:
This also enables us to implement straightforward receive props logic without rerendering:
```js
const Component = ({ bar }) => {
  const [foo, setFoo] = useState(0);
  if (bar) {
    setFoo(2);
  }
  /* ... */
}
```

Reply:
Perfect! That gives us the same behaviour as with the old component API. Very nice solution.